pkg/reconciler/sequence/sequence.go

```diff
@@ -224,3 +232,86 @@ func (r *Reconciler) trackAndFetchChannel(ctx context.Context, seq *v1.Sequence,
 	}
 	return obj, err
 }
+
+func (r *Reconciler) removeUnwantedChannels(ctx context.Context, channelResourceInterface dynamic.ResourceInterface, seq *v1.Sequence, wanted []*eventingduckv1.Channelable) error {
+	channelObjRef := corev1.ObjectReference{
+		Kind:       seq.Spec.ChannelTemplate.Kind,
+		APIVersion: seq.Spec.ChannelTemplate.APIVersion,
+	}
+
+	l, err := r.channelableTracker.ListerFor(channelObjRef)
+	if err != nil {
+		logging.FromContext(ctx).Errorw("Error getting lister for Channel", zap.Any("channelRef", channelObjRef), zap.Error(err))
+		return err
+	}
+
+	exists, err := l.ByNamespace(seq.GetNamespace()).List(labels.Everything())
+	if err != nil {
+		logging.FromContext(ctx).Errorw("Error listing Channels", zap.Any("namespace", seq.Namespace), zap.Any("channelRef", channelObjRef), zap.Error(err))
+		return err
+	}
+
+	for _, c := range exists {
+		ch, err := kmeta.DeletionHandlingAccessor(c)
+		if err != nil {
+			logging.FromContext(ctx).Errorw("Failed to get get channel", zap.Any("channel", c), zap.Error(err))
+			return err
+		}
+
+		if !ch.GetDeletionTimestamp().IsZero() ||
+			!metav1.IsControlledBy(ch, seq) {
+			continue
+		}
+
+		used := false
+		for _, cw := range wanted {
+			if cw.Name == ch.GetName() {
+				used = true
+				break
+			}
+		}
+
+		if !used {
+			err = channelResourceInterface.Delete(ctx, ch.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				logging.FromContext(ctx).Errorw("Failed to delete Channel", zap.Any("channel", ch), zap.Error(err))
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) removeUnwantedSubscriptions(ctx context.Context, seq *v1.Sequence, wanted []*eventingduckv1.Channelable) error {
+	subs, err := r.subscriptionLister.Subscriptions(seq.Namespace).List(labels.Everything())
+	if err != nil {
+		logging.FromContext(ctx).Errorw("Error listing Subscriptions", zap.Any("namespace", seq.Namespace), zap.Error(err))
+		return err
+	}
+
+	for _, sub := range subs {
+		if !sub.GetDeletionTimestamp().IsZero() ||
+			!metav1.IsControlledBy(sub, seq) {
+			continue
+		}
+
+		used := false
+		for _, ch := range wanted {
+			if sub.Spec.Channel.Name == ch.Name {
```

Comment:
I find this a bit surprising. I would expect `wanted` to be the list of expected subscriptions. I think the end result is the same but what you did for channel looks simpler. WDYT?

Reply:
I think you are absolutely right.
https://github.com/knative/eventing/pull/5718/commits/8ba440d97fc2f61018a43475a136752e490fff68